mqtt/mqtt-bridge/src/client.rs

```diff
@@ -32,55 +39,154 @@ impl ShutdownHandle {
+
+            let token_task = async {
+                match token_source {
+                    Some(ts) => ts.get(&expiry).await,
+                    None => Ok(None),
+                }
+            };
+
+            let (password, io) = future::try_join(token_task, io).await.map_err(|err| {
+                Error::new(ErrorKind::Other, format!("failed to connect: {}", err))
+            })?;
+
+            let stream: Pin<Box<dyn BridgeIo>> = Box::pin(io);
+            Ok((stream, password))
+        })
+    }
+
+    fn get_tls_source(connection_settings: &TcpConnection<SasTokenSource>) -> BridgeIoSourceFuture {
+        let address = connection_settings.address.clone();
+        let token_source = connection_settings.token_source.as_ref().cloned();
+        let trust_bundle_source = connection_settings.trust_bundle_source.clone();
 
         Box::pin(async move {
             let expiry = Utc::now() + chrono::Duration::minutes(DEFAULT_TOKEN_DURATION_MINS);
 
-            let password: Option<String> = if let Some(ts) = token_source {
-                ts.get(&expiry)
+            let server_root_certificate_task = async {
+                match trust_bundle_source {
+                    Some(source) => source.get_trust_bundle().await,
+                    None => Ok(None),
+                }
+            };
+
+            let token_task = async {
+                match token_source {
+                    Some(ts) => ts.get(&expiry).await,
+                    None => Ok(None),
+                }
+            };
+
+            let io = TcpStream::connect(address.clone());
+
+            let (server_root_certificate, password, stream) =
+                future::try_join3(server_root_certificate_task, token_task, io)
                     .await
-                    .map_err(|e| {
-                        error!("Failed to get token for connection {} {}", address, e);
-                        e
-                    })
-                    .ok()
+                    .map_err(|err| {
+                        Error::new(ErrorKind::Other, format!("failed to connect: {}", err))
+                    })?;
+
+            let config = SslConnector::builder(SslMethod::tls())
+                .map(|mut builder| {
+                    if let Some(trust_bundle) = server_root_certificate {
+                        X509::stack_from_pem(trust_bundle.as_bytes())
+                            .map(|mut certs| {
+                                while let Some(ca) = certs.pop() {
+                                    builder.cert_store_mut().add_cert(ca).ok();
+                                }
+                            })
+                            .ok();
+                    }
+
+                    builder.build()
+                })
+                .and_then(|conn| conn.configure())
+                .ok();
+
+            let hostname = address.split(':').next().unwrap_or(&address);
+
+            let res = if let Some(c) = config {
```

Comment:
nit: also `config.map(...)`?

Reply:
I couldn't use map because of the await, is there a way to use map with async?

Reply:
I believe you could just mark lambda as async.

Reply:
async closures are not in stable rust yet :(